soroban-env-common/src/env_val.rs

```diff
@@ -236,10 +236,13 @@ where
                 hi: 0,
                 lo: val.get_body(),
             })),
-            Tag::I128Small => Ok(ScVal::I128(Int128Parts {
-                hi: 0,
-                lo: val.get_signed_body() as u64,
-            })),
+            Tag::I128Small => {
+                let body = val.get_signed_body() as i128;
+                Ok(ScVal::I128(Int128Parts {
+                    hi: (body >> 64) as u64,
+                    lo: body as u64,
```

Comment:
is this the same thing as `(body << 64) >> 64`? i.e. it won't consider the top bits of `body` in the casting?

Reply:
Yeah `as u64` will truncate the more significant bits that don't fit.

Reply:
no, the first line sign-extends i64 to i128, so if get_signed_body returns a negative i64 (eg. 0xffff...)  then body will be a negative i128 (also 0xffff...) and then shifting it back down will produce -1_i64. We could also have put `hi: if body < 0 { -1 } else { 0 }` but this seemed more straightforward.

Reply:
yeah I was commenting specifically on line 243, the `lo:` part